File: libs/go/streamgraph.go
package stonecharts

import (
	"fmt"
	"strings"
)

func renderStreamgraphSVG(spec *ChartSpec) string {
	return renderCartesian(spec, "Streamgraph", "point", streamgraphMarks, false)
}

func streamgraphMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	K := len(f.spec.Series)
	N := f.n

	baseline := f.sgBaseline
	cumBottom := f.sgCumBottom
	cumTop := f.sgCumTop

	allVals := make([][]float64, K)
	for k := 0; k < K; k++ {
		vals := make([]float64, 0, N)
		for i, v := range f.spec.Series[k].Data {
			if i >= N {
				break
			}
			vals = append(vals, v)
		}
		allVals[k] = vals
	}

	for si := 0; si < K; si++ {
		s := f.spec.Series[si]
		st := f.styles[si]
		rawVals := allVals[si]

		topPts := make([][2]float64, len(rawVals))
		bottomPts := make([][2]float64, len(rawVals))
		for i := range rawVals {
			x := f.xpix(float64(i))
			topY := f.ypix(baseline[i] + cumTop[si][i])
			botY := f.ypix(baseline[i] + cumBottom[si][i])
			topPts[i] = [2]float64{x, topY}
			bottomPts[i] = [2]float64{x, botY}
		}

		fmt.Fprintf(p, `<g class="sc-series" data-series="%d">`, si)

		if len(topPts) > 0 {
			topD := areaTopPath(topPts, "", s.Curve)
			bottomRev := make([][2]float64, len(bottomPts))
			for i := range bottomPts {
				bottomRev[i] = bottomPts[len(bottomPts)-1-i]
			}
			bottomD := areaTopPath(bottomRev, "", s.Curve)
			if strings.HasPrefix(bottomD, "M") {
				bottomD = "L" + bottomD[1:]
			}
			ribbonD := topD + " " + bottomD + " Z"

			fill := st.fill
			fillOp := st.areaOp
			fmt.Fprintf(p,
				`<path class="sc-series-area" data-series="%d" d="%s" fill="%s"%s stroke="none"/>`,
				si, ribbonD, fill, fillOp)
		}

		if s.markerEnabled() {
			radius := s.markerRadius()
			radiusHover := radius + 2.5
			symbol := s.markerSymbol()
			for i, pt := range topPts {
				xlabel := fmt.Sprintf("%d", i)
				if i < len(f.cats) {
					xlabel = f.cats[i]
				}
				common := fmt.Sprintf(
					`class="sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="%s" data-r-hover="%s"`,
					si, esc(s.Name), esc(xlabel), esc(fmtNum(rawVals[i])), st.solid, fmtNum(radius), fmtNum(radiusHover))
				p.WriteString(markerSVG(symbol, pt[0], pt[1], radius, common, st.solid, f.theme.MarkerHalo, 1.0))
			}
		}
		p.WriteString(`</g>`)
	}
}

File: libs/python/stonecharts/charts/streamgraph.py
"""Streamgraph chart renderer: ChartSpec -> SVG string.

Stacked, filled area ribbons over a shared x-axis displaced off a floating
baseline (wiggle or silhouette). Rides the area renderer's path builder and
the shared cartesian frame with include_zero=False.
"""

from __future__ import annotations

from ..spec import Marker
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian
from .line import _marker, _path_d, _spline_d


def render_svg(spec) -> str:
    return render_cartesian(spec, "Streamgraph", "point", _streamgraph_marks,
                            include_zero=False)


def _streamgraph_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    K = len(fr.spec.series)
    N = fr.n

    baseline = fr.sg_baseline
    cum_bottom = fr.sg_cum_bottom
    cum_top = fr.sg_cum_top

    all_vals: list[list[float]] = []
    for s in fr.spec.series:
        all_vals.append([float(v) for v in s.data[:N]])

    theme = fr.theme
    for si in range(K):
        s = fr.spec.series[si]
        st = fr.styles[si]
        raw_vals = all_vals[si]

        top_pts = []
        bottom_pts = []
        for i in range(len(raw_vals)):
            x = fr.xpix(i)
            top_y = fr.ypix(baseline[i] + cum_top[si][i])
            bot_y = fr.ypix(baseline[i] + cum_bottom[si][i])
            top_pts.append((x, top_y))
            bottom_pts.append((x, bot_y))

        p.append(f'<g class="sc-series" data-series="{si}">')

        if top_pts:
            top_d = _spline_d(top_pts) if s.curve == "monotone" else _path_d(top_pts, None)
            bottom_rev = list(reversed(bottom_pts))
            bottom_d = _spline_d(bottom_rev) if s.curve == "monotone" else _path_d(bottom_rev, None)
            if bottom_d.startswith("M"):
                bottom_d = "L" + bottom_d[1:]
            ribbon_d = f"{top_d} {bottom_d} Z"

            fill = st.fill
            fill_op = st.area_op if st.area_op else ""
            p.append(
                f'<path class="sc-series-area" data-series="{si}" d="{ribbon_d}" '
                f'fill="{fill}"{fill_op} stroke="none"/>'
            )

        mk = s.marker or Marker()
        if mk.enabled and top_pts:
            radius = mk.radius
            radius_hover = radius + 2.5
            for i, (x, y) in enumerate(top_pts):
                xlabel = fr.cats[i] if i < len(fr.cats) else str(i)
                common = (
                    f'class="sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                    f'data-x="{esc(xlabel)}" data-y="{esc(fmt_num(raw_vals[i]))}" '
                    f'data-color="{st.solid}" data-r="{fmt_num(radius)}" '
                    f'data-r-hover="{fmt_num(radius_hover)}"'
                )
                p.append(_marker(mk.symbol, x, y, radius, common, st.solid, theme.marker_halo))
        p.append("</g>")
